concurrency: only count active waiter in totalAndMaxWaitDuration metric calculation

lyang24 opened the pull request:
concurrency: only count active waiter in totalAndMaxWaitDuration metric calculation.
   
The lock wait duration should only apply for the waiters that are actively waiting. This change 
has no effect on wating readers since all waiting readers are active. We are excluding all 
inactive locking requests' wait time contribution to the total and max waiting duration.

Fixes: #108683
Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108797)
<!-- Reviewable:end -->

arulajmani reviewed (commented):
Thanks for picking this up and sending out the patch! The code change itself looks good -- I have a few minor comments around adding a targeted test and formatting of the commit message, but other than that, this is great!

> Reviewed 1 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-:-NdGO-YYDpsADOZhgDOm:b99cj1b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-NdGMm987r4M_KoRSSRC):*
The title of the PR should also be included in the commit message. There are a few guidelines about CRDB specific commit messages in this confluence doc: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#CockroachDB-specific-guidelines

Additionally, we're only changing the `concurrency` package here, so let's prefix the change with `concurrency:` instead of `kv:`
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1777 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-NdGNIBR-zE18R0g39Um:-NdGNIBR-zE18R0g39Un:bf3vd0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4d1f8165de078504b61822b44f9bfa143d3f2ba/pkg/kv/kvserver/concurrency/lock_table.go#L1777)):*
> ```Go
> 
> // Returns the total amount of time all waiters in the queues of
> // readers and writers have been waiting on the lock.
> ```

Should we update this comment to reference active waiters as well?
___
*[`pkg/kv/kvserver/concurrency/testdata/lock_table/basic` line 288 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-NdGN_4_CY-EhHejpUvg:-NdGN_4_CY-EhHejpUvh:b-oiyxcy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4d1f8165de078504b61822b44f9bfa143d3f2ba/pkg/kv/kvserver/concurrency/testdata/lock_table/basic#L288)):*
> ```
> waiters: 3
> waitingreaders: 0
> waitingwriters: 3
> ```

Should we also add another test in this file that shows the targeted effect of your change? 


<!-- Sent from Reviewable.io -->

Commit:
concurre:sency: only count active waiter in totalAndMaxWaitDuration metric calculation.

The lock wait duration should only apply for the waiters that are actively waiting. This change
has no effect on wating readers since all waiting readers are active. We are excluding all
inactive waiting locking requests' wait time contribution to the total and max waiting duration.

Fixes #108683

Release note: None.

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108797)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-NdGMm987r4M_KoRSSRC):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

The title of the PR should also be included in the commit message. There are a few guidelines about CRDB specific commit messages in this confluence doc: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#CockroachDB-specific-guidelines

Additionally, we're only changing the `concurrency` package here, so let's prefix the change with `concurrency:` instead of `kv:`
</blockquote></details>

thanks fixed the commit message and pr title
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1777 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-NdGNIBR-zE18R0g39Um:-NdS72xR4zE3GibZEWzB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4d1f8165de078504b61822b44f9bfa143d3f2ba/pkg/kv/kvserver/concurrency/lock_table.go#L1777)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we update this comment to reference active waiters as well?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/concurrency/testdata/lock_table/basic` line 288 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-NdGN_4_CY-EhHejpUvg:-NdS74dg6HGE046Z0zOJ:bl3miq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4d1f8165de078504b61822b44f9bfa143d3f2ba/pkg/kv/kvserver/concurrency/testdata/lock_table/basic#L288)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we also add another test in this file that shows the targeted effect of your change?
</blockquote></details>

added a test targeting metrics exclude inactive waiters are the very end of the file


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
One minor request re: a comment, but once that's addressed, this should be good to merge. 

> Reviewed 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-:-Nd_y6zF7qrzWNzlNwBN:b-ipoimo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-Nd_wqIC3BJMkF3fh6wu):*
nit: "that are actively waiting"
___
*[`-- commits` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-Nd_wyLfCQxmAyWvuYS8):*
We've recently change the verbiage here to apply to "locking requests" as opposed to writers. That's because we've introduced shared locking requests, which are locking requests that do not perform writes. Should we update this to say "exclusive all inactive waiting locking requests' wait time" instead?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1777 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-NdGNIBR-zE18R0g39Um:-Nd_xPCXAhMPSOBIuIwn:b-khfuxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4d1f8165de078504b61822b44f9bfa143d3f2ba/pkg/kv/kvserver/concurrency/lock_table.go#L1777)):*
<details><summary><i>Previously, lyang24 (Lanqing Yang) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

s/active locking requests/locking requests/g
___
*[`pkg/kv/kvserver/concurrency/testdata/lock_table/basic` line 288 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-NdGN_4_CY-EhHejpUvg:-Nd_y2jiE5XElvKd8wNJ:b-r60t5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4d1f8165de078504b61822b44f9bfa143d3f2ba/pkg/kv/kvserver/concurrency/testdata/lock_table/basic#L288)):*
<details><summary><i>Previously, lyang24 (Lanqing Yang) wrote…</i></summary><blockquote>

added a test targeting metrics exclude inactive waiters are the very end of the file
</blockquote></details>

Nice, thanks for adding the test!


<!-- Sent from Reviewable.io -->

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108797)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-Nd_wqIC3BJMkF3fh6wu):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: "that are actively waiting"
</blockquote></details>

fixed
___
*[`-- commits` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-Nd_wyLfCQxmAyWvuYS8):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We've recently change the verbiage here to apply to "locking requests" as opposed to writers. That's because we've introduced shared locking requests, which are locking requests that do not perform writes. Should we update this to say "exclusive all inactive waiting locking requests' wait time" instead?
</blockquote></details>

great fixed the wording of commit message
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1777 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-NdGNIBR-zE18R0g39Um:-NdaYgoX5WqMONXMIGlh:b5cwun2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4d1f8165de078504b61822b44f9bfa143d3f2ba/pkg/kv/kvserver/concurrency/lock_table.go#L1777)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

s/active locking requests/locking requests/g
</blockquote></details>

replaced completed in the scope of this comment.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> This should be good to go!

bors r+

> Reviewed 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108797#-:-NdazaKGD094mqbqT3FA:b-bszhu1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lyang24)



<!-- Sent from Reviewable.io -->
